src/test/java/org/jenkinsci/plugins/workflow/cps/global/GrapeTest.java

```diff
@@ -0,0 +1,174 @@
+                story.j.assertLogContains("net.nowhere", story.j.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0)));
+            }
+        });
+    }
+
+    @Ignore("TODO ditto")
+    @Test public void nonexistentImport() throws Exception {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                FileUtils.write(new File(repo.workspace, "src/pkg/X.groovy"),
+                    "package pkg\n" +
+                    "@Grab('commons-primitives:commons-primitives:1.0')\n" +
+                    "import net.nowhere.Nonexistent\n" +
+                    "static def run() {new Nonexistent()}");
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition("pkg.X.run()", true));
+                story.j.assertLogContains("net.nowhere.Nonexistent", story.j.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0)));
+            }
+        });
+    }
+
+    // TODO test alternate Maven repositories
+
+    @Ignore("TODO MissingMethodException: No signature of method: static com.google.common.base.CharMatcher.whitespace() is applicable for argument types: () values: []")
+    @Test public void overrideCoreLibraries() throws Exception {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                FileUtils.write(new File(repo.workspace, "src/pkg/Strings.groovy"),
+                    "package pkg\n" +
+                    "@Grab('com.google.guava:guava:19.0')\n" + // 11.0.1 from core has only WHITESPACE constant
+                    "import com.google.common.base.CharMatcher\n" +
+                    "static def hasWhitespace(text) {\n" +
+                    "  CharMatcher.whitespace().matchesAnyOf(text)\n" +
+                    "}");
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition("echo(/checking ${pkg.Strings.hasWhitespace('hello world')}/)", true));
+                story.j.assertLogContains("checking true", story.j.buildAndAssertSuccess(p));
+            }
+        });
+    }
+
+    @Ignore("TODO works in Jenkins 2 / Groovy 2; in 1, fails with (concealed): unable to resolve class org.jenkinsci.plugins.docker.workflow.Docker")
+    @Test public void useSource() throws Exception {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                FileUtils.write(new File(repo.workspace, "src/pkg/Dokker.groovy"),
+                    "package pkg\n" +
+                    "@Grapes([@Grab('org.jenkins-ci.plugins:docker-workflow:1.7'), @Grab('org.jenkins-ci.plugins:docker-commons:1.3.1')])\n" +
+                    "import org.jenkinsci.plugins.docker.workflow.Docker\n" +
+                    "static def stuff(script, body) {\n" +
+                    "  new Docker(script).node {body()}\n" +
```

Comment:
Not a realistic use case, but I struggled to find a JAR in Central containing runnable Groovy source files—especially one accepting closure arguments (used here to prove that the library gets CPS-transformed).